### PR TITLE
[MAINTENANCE] SQLAlchemy 2 typing

### PR DESCRIPTION
**Pull request by Kilo59**
Update the type-checking step to install sqlalchemy 2.

We currently support both sqlalchemy 1 + 2 but sqlalchemy 2 is the only version that exports types.

Too many issues to fix immediately so type-ignores have been added for existing violations.

### CODE REVIEW

File: contrib/cli/requirements.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Click>=7.1.2         # CLI tooling
 cookiecutter==2.1.1  # Project templating
-mypy==1.10.0            # Type checker
+mypy==1.10.1.0            # Type checker
```
Comment (Kilo59) [inline contrib/cli/requirements.txt]:
```suggestion
mypy==1.10.1          # Type checker
```

File: constraints-dev.txt [hunk 2]
```diff
@@ -29,3 +29,4 @@
 # protobuf 3.19.5, 3.19.6, and 3.20.2 were released in September 2022
 # https://pypi.org/project/protobuf/#history
 protobuf!=3.20.0,!=3.20.1,<3.21
+setuptools>=70.0.0,<71.0.0 # lower bound required for python 3.12, upper bound required for how we use pkg_resources in scripts
```
Comment (Kilo59) [inline constraints-dev.txt]:
```suggestion
```

File: reqs/requirements-dev-bigquery.txt [hunk 3]
```diff
@@ -6,4 +6,4 @@ google-cloud-secret-manager>=1.0.0
 # needed to prevent excessive pip backtracking on python 3.11
 google-cloud-storage>=2.10.0;python_version >= '3.11'
 google-cloud-storage>=1.28.0;python_version < '3.11'
-sqlalchemy-bigquery>=1.3.0
+sqlalchemy-bigquery>=1.11.0
```
Comment (Kilo59) [inline reqs/requirements-dev-bigquery.txt]:
```suggestion
sqlalchemy-bigquery>=1.3.0
```

File: reqs/requirements-dev-sqlalchemy2.txt [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Do not `pip install -r` this file locally or in CI (outside of StaticTypeCheck).
+# It only exists to be included in requirements-dev.txt, which should only be used in the Dockerfile.tests file.
+# Do not add any dependencies here that are not compatible with sqlalchemy>=2.0
+--requirement requirements-dev-bigquery.txt
```
Comment (Kilo59) [inline reqs/requirements-dev-sqlalchemy2.txt]:
```suggestion
sqlalchemy-bigquery>=1.11.0
```

File: setup.py [hunk 5]
```diff
@@ -30,14 +30,13 @@ def get_extras_require():
...
     sqla1x_only_keys = (
-        "bigquery",  # https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/setup.py
         "clickhouse",  # https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/master/setup.py
         "redshift",  # https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/blob/main/setup.py
-        "snowflake",  # https://github.com/snowflakedb/snowflake-sqlalchemy/blob/main/setup.cfg
         "teradata",  # https://pypi.org/project/teradatasqlalchemy   https://support.teradata.com/knowledge?id=kb_article_view&sys_kb_id=a5a869149729251ced863fe3f153af27
     )
     sqla_keys = (
         "athena",  # https://github.com/laughingman7743/PyAthena/blob/master/pyproject.toml
+        "bigquery",  # https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/setup.py
```
Comment (Kilo59) [inline setup.py]:
```suggestion
```

File: setup.py [hunk 6]
```diff
@@ -30,14 +30,13 @@ def get_extras_require():
         "s3": "boto",
     }
     sqla1x_only_keys = (
-        "bigquery",  # https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/setup.py
         "clickhouse",  # https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/master/setup.py
```
Comment (Kilo59) [inline setup.py:34]:
```suggestion
        "bigquery",  # https://github.com/googleapis/python-bigquery-sqlalchemy/blob/main/setup.py
        "clickhouse",  # https://github.com/xzkostyan/clickhouse-sqlalchemy/blob/master/setup.py
```

File: great_expectations/expectations/metrics/column_aggregate_metrics/column_quantile_values.py [hunk 7]
```diff
@@ -209,13 +209,14 @@ def _get_column_quantiles_bigquery(
 ) -> list:
     # BigQuery does not support "WITHIN", so we need a special case for it
     selects: list[sqlalchemy.WithinGroup] = [
-        sa.func.percentile_disc(column, quantile).over() for quantile in quantiles
+        sa.func.percentile_disc(column, quantile).over()
+        for quantile in quantiles  # type: ignore[misc]
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/column_aggregate_metrics/column_quantile_values.py:213]:
```suggestion
        for quantile in quantiles
```

File: great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py [hunk 8]
```diff
@@ -543,27 +543,27 @@ def _sqlalchemy_map_condition_index(  # noqa: C901 - too complex
...
         unexpected_condition
     )
 
     if not _is_sqlalchemy_metric_selectable(map_metric_provider=cls):
-        domain_records_as_selectable = get_sqlalchemy_selectable(domain_records_as_selectable)
+        domain_records_as_selectable = get_sqlalchemy_selectable(domain_records_as_selectable)  # type: ignore[arg-type]
 
     # since SQL tables can be **very** large, truncate query_result values at 20, or at `partial_unexpected_count`  # noqa: E501
-    final_query: sa.select = unexpected_condition_query_with_selected_columns.select_from(
+    final_query: sa.select = unexpected_condition_query_with_selected_columns.select_from(  # type: ignore[valid-type]
```
Comment (billdirks) [inline great_expectations/expectations/metrics/map_metric_provider/map_condition_auxilliary_methods.py]:
```suggestion
    final_query: sa.select = unexpected_condition_query_with_selected_columns.select_from(  # type: ignore[valid-type,attr-defined]
```

File: great_expectations/expectations/metrics/table_metrics/table_head.py [hunk 9]
```diff
@@ -75,10 +75,10 @@ def _sqlalchemy(
         if metric_value_kwargs["fetch_all"]:
             limit = None
 
-        selectable = sa.select("*").select_from(selectable).limit(limit).selectable
+        selectable = sa.select("*").select_from(selectable).limit(limit).selectable  # type: ignore[assignment]
```
Comment (billdirks) [inline great_expectations/expectations/metrics/table_metrics/table_head.py]:
```suggestion
        selectable = sa.select("*").select_from(selectable).limit(limit).selectable  # type: ignore[assignment,arg-type]
```

File: great_expectations/expectations/metrics/column_aggregate_metrics/column_quantile_values.py (hunk 7 above)
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/column_aggregate_metrics/column_quantile_values.py]:
```suggestion
        sa.func.percentile_disc(column, quantile).over()  # type: ignore[misc]
        for quantile in quantiles
```

File: great_expectations/expectations/metrics/query_metrics/query_table.py [hunk 10]
```diff
@@ -56,7 +56,7 @@ def _sqlalchemy(
         else:
             query = query.format(batch=f"({selectable})")
 
-        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()
+        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment]
```
Comment (billdirks) [inline great_expectations/expectations/metrics/query_metrics/query_table.py]:
```suggestion
        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment,arg-type]
```

File: great_expectations/expectations/metrics/query_metrics/query_multiple_columns.py [hunk 11]
```diff
@@ -76,7 +76,7 @@ def _sqlalchemy(
                 batch=f"({selectable})",
             )
 
-        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()
+        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment]
```
Comment (billdirks) [inline great_expectations/expectations/metrics/query_metrics/query_multiple_columns.py]:
```suggestion
        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment,arg-type]
```

File: great_expectations/expectations/metrics/query_metrics/query_column.py [hunk 12]
```diff
@@ -60,7 +60,7 @@ def _sqlalchemy(
         else:
             query = query.format(col=column, batch=f"({selectable})")
 
-        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()
+        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment]
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/query_metrics/query_column.py]:
```suggestion
        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment,arg-type]
```

File: great_expectations/expectations/metrics/query_metrics/query_column_pair.py [hunk 13]
```diff
@@ -63,7 +63,7 @@ def _sqlalchemy(
         else:
             query = query.format(column_A=column_A, column_B=column_B, batch=f"({selectable})")
 
-        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()
+        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment]
```
Comment (billdirks) [inline great_expectations/expectations/metrics/query_metrics/query_column_pair.py]:
```suggestion
        result: List[sqlalchemy.Row] = execution_engine.execute_query(sa.text(query)).fetchall()  # type: ignore[assignment,arg-type]
```

File: great_expectations/compatibility/databricks.py [hunk 14]
```diff
@@ -5,6 +5,6 @@
 )
 
 try:
-    from databricks import connect  # type: ignore[import-not-found]
+    from databricks import connect  # type: ignore[import-not-found]  # type: ignore[import-untyped]
```
Comment (Kilo59) [inline great_expectations/compatibility/databricks.py]:
```suggestion
    from databricks import connect  # type: ignore[import-untyped]
```

File: great_expectations/expectations/metrics/util.py [hunk 15]
```diff
@@ -519,7 +519,7 @@ def column_reflection_fallback(  # noqa: C901, PLR0912, PLR0915
                     columns_table_query.c.column_precision,
                 )
                 .select_from(
-                    tables_table_query.join(
+                    tables_table_query.join(  # type: ignore[call-arg]
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/util.py]:
```suggestion
                    tables_table_query.join(  # type: ignore[call-arg,arg-type]
```

File: great_expectations/expectations/metrics/util.py [hunk 16]
```diff
@@ -593,14 +593,14 @@ def column_reflection_fallback(  # noqa: C901, PLR0912, PLR0915
...
-                sa.select(
+                sa.select(  # type: ignore[assignment]
                     tables_table_query.c.schema_name,
                     tables_table_query.c.table_name,
                     columns_table_query.c.column_name,
                     columns_table_query.c.column_data_type,
                 )
                 .select_from(
-                    tables_table_query.join(
+                    tables_table_query.join(  # type: ignore[call-arg]
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/util.py]:
```suggestion
                    tables_table_query.join(  # type: ignore[call-arg,arg-type]
```

File: great_expectations/expectations/metrics/util.py [hunk 17]
```diff
@@ -635,21 +635,21 @@ def column_reflection_fallback(  # noqa: C901, PLR0912, PLR0915
...
                 query: sqlalchemy.TextClause = selectable
-            elif sqlalchemy.Table and isinstance(selectable, sqlalchemy.Table):
+            elif sqlalchemy.Table and isinstance(selectable, sqlalchemy.Table):  # type: ignore[truthy-function]
                 query = sa.select(sa.text("*")).select_from(selectable).limit(1)
             else:  # noqa: PLR5501
                 # noinspection PyUnresolvedReferences
                 if dialect.name.lower() == GXSqlDialect.REDSHIFT:
                     # Redshift needs temp tables to be declared as text
-                    query = sa.select(sa.text("*")).select_from(sa.text(selectable)).limit(1)
+                    query = sa.select(sa.text("*")).select_from(sa.text(selectable)).limit(1)  # type: ignore[assignment]
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/util.py]:
```suggestion
                    query = sa.select(sa.text("*")).select_from(sa.text(selectable)).limit(1)  # type: ignore[assignment,arg-type]
```

File: great_expectations/expectations/metrics/util.py [hunk 18]
```diff
@@ -635,21 +635,21 @@ def column_reflection_fallback(  # noqa: C901, PLR0912, PLR0915
...
                 else:
-                    query = sa.select(sa.text("*")).select_from(sa.text(selectable)).limit(1)
+                    query = sa.select(sa.text("*")).select_from(sa.text(selectable)).limit(1)  # type: ignore[assignment]
```
Comment (anthonyburdi) [inline great_expectations/expectations/metrics/util.py]:
```suggestion
                    query = sa.select(sa.text("*")).select_from(sa.text(selectable)).limit(1)  # type: ignore[assignment,arg-type]
```

**Review by anthonyburdi — COMMENTED** [review]
Changes to great_expectations/expectations/metrics/util.py

File: reqs/requirements-dev-lite.txt [hunk 19]
```diff
@@ -20,7 +20,7 @@ pytest-timeout>=2.3.1
 pytest-xdist>=3.6.1
 requirements-parser>=0.9.0
 responses>=0.23.1 # requests mocking
-setuptools>=70.0.0 # required for python 3.12
+setuptools>=70.0.0 # lower bound required for python 3.12
```
Comment (Kilo59) [inline reqs/requirements-dev-lite.txt]:
```suggestion
setuptools>=70.0.0 # required for python 3.12
```

File: requirements-types.txt [hunk 20]
```diff
@@ -1,9 +1,10 @@
 --requirement requirements.txt
+--requirement reqs/requirements-dev-lite.txt
```
Comment (Kilo59) [inline requirements-types.txt:2]:
```suggestion
```

Comment (Kilo59) [inline requirements-types.txt:3]: `dev-lite` was a transitive dependency of this CI step before. Now I'm making it explicit.

**Review by anthonyburdi — APPROVED** [review]

**Pull request merged**